cpu/atmega_common/periph/rtt.c

```diff
@@ -0,0 +1,293 @@
+    TCCR2A = 0;
+
+    /* 32768Hz / 1024 = 32 ticks per second */
+    TCCR2B = (1 << CS22) | (1 << CS21) | (1 << CS20);
+
+    /* Wait until not busy anymore */
+    DEBUG("RTT waits until ASSR not busy\n");
+    _asynch_wait();
+
+    /* Clear interrupt flags */
+    /* Oddly, this is done by writing ones; see datasheet */
+    TIFR2 = (1 << OCF2B) | (1 << OCF2A) | (1 << TOV2);
+
+    /* Enable 8-bit overflow interrupt */
+    TIMSK2 |= (1 << TOIE2);
+
+    DEBUG("RTT initialized\n");
+}
+
+void rtt_set_overflow_cb(rtt_cb_t cb, void *arg)
+{
+    /* Interrupt safe order of assignment */
+    rtt_state.overflow_cb = NULL;
+    rtt_state.overflow_arg = arg;
+    rtt_state.overflow_cb = cb;
+}
+
+void rtt_clear_overflow_cb(void)
+{
+    /* Interrupt safe order of assignment */
+    rtt_state.overflow_cb = NULL;
+    rtt_state.overflow_arg = NULL;
+}
+
+uint32_t rtt_get_counter(void)
+{
+    /* Make sure it is safe to read TCNT2, in case we just woke up */
+    DEBUG("RTT sleeps until safe to read TCNT2\n");
+    TCCR2A = 0;
+    _asynch_wait();
+
+    return (((uint32_t)rtt_state.ext_cnt << 8) | (uint32_t)TCNT2);
+}
+
+void rtt_set_counter(uint32_t counter)
+{
+    /* Wait until not busy anymore (should be immediate) */
+    DEBUG("RTT sleeps until safe to write TCNT2\n");
+    _asynch_wait();
+
+    rtt_state.ext_cnt = (uint16_t)(counter >> 8);
+    TCNT2 = (uint8_t)counter;
+
+    DEBUG("RTT set counter TCNT2: %" PRIu8 ", ext_cnt: %" PRIu16 "\n",
+          TCNT2, rtt_state.ext_cnt);
+}
+
+void rtt_set_alarm(uint32_t alarm, rtt_cb_t cb, void *arg)
+{
+    uint32_t now;
+
+    /* Disable alarm interrupt */
+    TIMSK2 &= ~(1 << OCIE2A);
+    rtt_state.alarm_cb = NULL;
+
+    /* Make sure it is safe to read TCNT2, in case we just woke up, and */
+    /* safe to write OCR2B (in case it was busy) */
+    DEBUG("RTC sleeps until safe read TCNT2 and to write OCR2B\n");
+    TCCR2A = 0;
+    _asynch_wait();
+
+    now = ((uint32_t)rtt_state.ext_cnt << 8) | TCNT2;
+
+    if (alarm < now) {
+        DEBUG("RTC alarm set in the past. Time: %" PRIu32 " seconds, alarm: %"
+              PRIu32 "\n", now, alarm);
+        cb(arg);
+        return;
+    }
+
+    /* Set the alarm value */
+    rtt_state.ext_comp = (uint16_t)(alarm >> 8);
+    OCR2A = (uint8_t)alarm;
+
+    DEBUG("RTT set alarm TCNT2: %" PRIu8 ", OCR2A: %" PRIu8 "\n", TCNT2, OCR2A);
+
+    /* Interrupt safe order of assignment */
+    rtt_state.alarm_arg = arg;
+    rtt_state.alarm_cb = cb;
+
+    /* Enable alarm interrupt only if it will trigger before overflow */
+    if (rtt_state.ext_comp <= rtt_state.ext_cnt) {
+        /* Clear interrupt flag */
+        TIFR2 = (1 << OCF2A);
+
+        /* Enable interrupt */
+        TIMSK2 |= (1 << OCIE2A);
+
+        DEBUG("RTT alarm interrupt active\n");
+    }
+    else {
+        DEBUG("RTT alarm interrupt not active\n");
+    }
+}
+
+uint32_t rtt_get_alarm(void)
+{
+    return (((uint32_t)rtt_state.ext_comp << 8) | (uint32_t)OCR2A);
```

Comment:
Somehow my inline comments vanished.

Here you are accessing `rtt_state_ext_comp`, which is 16 bit of RAM. On an 8-bit CPU this access will be two instructions. Afterwards you'll access `OCR2A`, which will add more instructions. If between those instructions an interrupt kicks in and a different thread gets scheduled `rtt_state.ext_comp` may be changed. As a result the two 8 bit chunks of `rtt_state.ext_comp` accessed might belong to different states of `rtt_state.ext_comp` and might combined have a result that was neither correct before nor after the interrupt. This is racy and `volatile` is not helping here. Also C11 atomics will not help (but at least fail on compilation), as there is no way to read both `rtt_sate.ext_comp` and `OCR2A` in one instant.

Reply:
Out of scope for the reason why volatile is employed here. Consistency is only guaranteed for multiple threads if they are only reading. Multi-thread read/write consistency is not guaranteed unless global interrupts are disabled before making the API calls. However, multi-thread writes are in general undesirable for this API as you cannot e.g. set an alarm for each thread anyway.

Your critique applies to e.g. `rtt_get_counter` with respect to timer events, so I will work on fixing that.